### ISSUE TITLE
Code explanation

### ISSUE THREAD

**Issue by lainconn** [CV-Magician/MMM-CLIP#3]
Could you please explain: Why are you taking <code>text_features[0]</code> in <code>CustomDataset.get_image_features</code> in main.py file? Does not it resolve in taking the first prompt embeddings, in your case "A photo of motocycle"?

**Comment by YesianRohn** [comment]
This task is a multi-label classification, naturally requiring all the prompt words, and then obtaining their features. Of course, this is based on intuition, without experimental verification.

**Comment by lainconn** [comment]
I understand the need to skim through all possible classes, that's why you use <code>model.encode_text(text)</code> in the first place, but I am confused by indexing text_features. Does not it conclude in extracting only one text embedding of all?

Here is the code for simplicity, taken from `main.py`:

```python
def get_image_features(self, idx):
    global text_features
    item = self.data[idx]
    image_name = item["Image_Name"]
    image_path = os.path.join(self.image_folder, image_name)
    image = preprocess(Image.open(image_path)).unsqueeze(0).to(device)

    with torch.no_grad():
        image_features = model.encode_image(image)
        text_features = model.encode_text(text)
    features = torch.cat([image_features[0], text_features[0]], dim=-1)
    return features

**Comment by YesianRohn** [comment]
I think I understand what you mean.  The code `torch. cat([image_features[0], text_features[0]], dim=-1)` is not actually taking the first image or text;  it seems that the initial CLIP result list has nested redundancy, which is related to the way CLIP is used in the code.

**Comment by lainconn** [comment]
Actually, you are right if we speak about `image_features`, which is indeed indexed by `batch_size` (the shape of `image_features` is (batch_size, n_dim)), but the same does not apply to `text_features`, which have shape of (batch_size, n_dim), in your case `batch_size` are the number of text prompts. So in regard to `text_features[0]`, you just take the first prompt's embeddings.

**Comment by YesianRohn** [comment]
If that's the case, then you should be right, and I made a mistake here.
